### DISCUSSION TITLE
User-Mode File System

### DISCUSSION THREAD

**Discussion by TonyValenti** [dokan-dev/dokan-dotnet#329] [Q&A]
Hi!  I'm looking to build a file system layer on top of some JSON web APIs.  I'm thinking about Dokan but I have a few questions:

1.  I'm using .NET 7.0.  Will that cause any issues?

2.  I'd like my app to be able to run self-contained without having to actually be installed on the system.

2.1 Do the [dokan](https://github.com/dokan-dev/dokany) drivers have to be installed separately or are they included in the NUGET package?

2.2 Are non-administrators able to create Dokan drives?  I'd like my app to be able to work for non-administrator and non-elevated users.

3.  Looking at the interfaces, it looks like Dokan expects everything to be a sync operation.  Our code is heavily Async though.  Is there a set of Async interfaces or such that we could use?  Or is there any guidance on how to wrap a Task in a dokan-friendly way?

Thanks so much!

**Comment by Liryna** [comment]
Hi @TonyValenti ,

> I'm using .NET 7.0. Will that cause any issues?

That should be fine.

> I'd like my app to be able to run self-contained without having to actually be installed on the system.

Dokan or any other solution creating a virtual drive will need to install a driver on the system.

> 2.1 Do the [dokan](https://github.com/dokan-dev/dokany) drivers have to be installed separately or are they included in the NUGET package?

It needs to be installed separately. It contains the driver and the native library used by the Nuget package.

> 2.2 Are non-administrators able to create Dokan drives? I'd like my app to be able to work for non-administrator and non-elevated users.

Yes

> Looking at the interfaces, it looks like Dokan expects everything to be a sync operation. Our code is heavily Async though. Is there a set of Async interfaces or such that we could use? Or is there any guidance on how to wrap a Task in a dokan-friendly way?

Unfortunately no, there is a beginning of making the API async [here](https://github.com/dokan-dev/dokany/pull/307) but it requires some work to merge it.
That said, the API is running in Dokan threads which you can "block" for the time you need while keeping the request alive with `TryResetTimeout`.

**Reply by TonyValenti** [reply]
Thank you!

It sounds like the README file might need updating because it say:
![image](https://user-images.githubusercontent.com/13609558/231450886-5ae3382f-0342-44d1-bb67-df7687097d55.png)

**Reply by Liryna** [reply]
Good catch! Thanks 😄 Done.

**Comment by TonyValenti** [comment]
Also, in the mirror example, what is the difference between the Safe and Unsafe mirrors?

**Reply by Olof-Lagerkvist** [reply]
> Also, in the mirror example, what is the difference between the Safe and Unsafe mirrors?

It is a legacy concept from older .NET runtimes where there were very different ways to handle managed memory such as arrays or to handle unmanaged memory by pointers to and from unmanaged code. Nowadays, both are best done with `Span<byte>` and similar concepts in similar ways, but before that was available you would practically need two slightly different code paths throughout your libraries for API calls that expected managed memory vs unmanaged memory.

Since Dokan implementations receive unmanaged pointers for I/O buffers from an unmanaged dll, those buffers typically needed to be copied to and from temporary managed arrays to make it possible to send the requests on to manged calls like `Read` and `Write` on `Stream` objects and similar. But if the backend actually used unmanaged pointers anyway, you could avoid some unnecessary buffer copying etc by implementing the unsafe versions of the interface instead to receive the original unmanaged pointer directly and then use that one in your backend calls as well.
